File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    
    public void transpose(int[][] matrix){
        for(int i = 0;i<matrix.length;i++){
            for (int j = 0;j<matrix[0].length;j++){
                if(i<j){
                    int temp = matrix[i][j];
                    matrix[i][j] = matrix[j][i];
                    matrix[j][i] = temp;
                }
            }
        }
    }
    
   public void reverse(int[][] matrix) {
        for (int i = 0; i < matrix.length; i++) {
            int left = 0;
            int right = matrix[i].length - 1;
            while (left < right) {
                int temp = matrix[i][left];
                matrix[i][left] = matrix[i][right];
                matrix[i][right] = temp;
                left++;
                right--;
            }
        }
    }
    public void rotate(int[][] matrix) {
        
        this.transpose(matrix);
        this.reverse(matrix);
        
    }
}

File: 0048-rotate-image/0048-rotate-image.py
class Solution(object):
    def transpose(self, matrix):
        for i in range(len(matrix)):
            for j in range(len(matrix[0])):
                if i!=j and i<j:
                    matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
                    
    def reverse(self, matrix):
        for i in range(len(matrix)):
            matrix[i].reverse()
    
    #2. Reverse the rows 
    def rotate(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        
        self.transpose(matrix)
        self.reverse(matrix)
        
        
        
        